# Objet : Analyse des relations entre champs des données IRVE

## Contexte
- clarification du rôle des modèles de données dans les jeux de données (cf mise à jour récente des [guides data.gouv](https://guides.etalab.gouv.fr/qualite/documenter-les-donnees/))
- intégration d'une propriété "relationship" dans les schémas de données ([issue TableSchema](https://github.com/frictionlessdata/specs/issues/803) en cours de validation)
- création d'outils de contrôle des relations entre champs des jeux de données tabulaires (cf usage ci-dessous)

## Objectifs
- valider sur un cas réel l'utilisation d'un modèle de données en complément d'un schéma de données
- identifier les apports que pourraient avoir les contrôles de validation des relations entre champs

## Résultats
- la formalisation d'un modèle de données facilite la compréhension des données et des relations entre champs
- l'outil de contrôle permet d'améliorer significativement la qualité des données par la détection d'incohérences de relations
- l'analyse a posteriori permet de (re)trouver la logique de dépendance entre les colonnes qui minimise les incohérences
- les incohérences détectées sur le jeu de données IRVE restent faibles (inférieures à 2% des point de charge documentés - voir chapitre 4)

## Suite à donner
- Mettre à jour, valider et publier le modèle de données IRVE
- Définir les contrôles supplémentaires à intégrer pour toutes nouvelles données ainsi que pour le jeu complet
- Mettre en oeuvre les outils de contrôle

## Evolutions possibles 
- Ajouter dans les guides d'Etalab un guide pour les modèles de données 
- Intégrer dans les schémas de données la propriété "relationship" en cours de validation,
- Définir un indicateur qui mesure l'écart (existant / attendu) des relations entre champs

## Sommaire
*(active link on jupyter Notebook or Nbviewer)*
- [1 - modèle de données](#1---modèle-de-données)
- [2 - Initialisation](#2---Initialisation)
- [3 - Analyse des relations entre champs](#3---Analyse-des-relations-entre-champs)
- [4 - Contrôle des relations](#4---Contrôle-des-relations)
- [5 - Contrôle du schéma de données](#5---Contrôle-du-schéma-de-données)
- [6 - Exemples d'analyse de relations](#6---Exemples-d'analyse-de-relations)
- [Annexe : Initialisation du modèle de données 'mermaid'](#Annexe-:-Initialisation-du-modèle-de-données-'mermaid')

Ce Notebook peut être consulté sur [nbviewer](http://nbviewer.org/github/loco-philippe/Environmental-Sensing/tree/main/python/Validation/irve)

données utilisées : https://www.data.gouv.fr/fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/    
fichier : "*consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv*"
_________________

# 1 - modèle de données

Le modèle de données proposé ci-dessous est construit sur la base du schéma de données mis à disposition et du contenu du jeu de données.    
Il est à consolider en fonction de l'expertise des concepteurs et réutilisateurs (voir [guide méthodologique](https://github.com/loco-philippe/Environmental-Sensing/blob/main/property_relationship/FR_methodology.ipynb)).

*Notation:*
- *M : Mandatory - documentation obligatoire*
- *PK : Primary Key - identifiant unique de l'entité*
- *Root : champ associé à une ligne du tableau*

In [3]:
# la variable 'irve' est définie en fin de Notebook
mermaid(irve)

------
# 2 - Initialisation
## initialisation des données
- lecture du fichier issu de l'api

In [152]:
from collections import Counter
from observation import Ilist, Iindex, Analysis
import pandas as pd
res = {}
chemin = 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/main/python/Validation/irve/'
res['date'] = '2023-03-03'
file = 'consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv'
irve = pd.read_csv(chemin + file, sep=',', low_memory=False)
res['len_irve'] = len(irve)
print('irve : ', res['len_irve']) 

irve :  51276


In [153]:
relations = ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
           'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance',  
           'date_maj', 'last_modified']
champs = ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
           'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance', 
           'puissance_nominale', 'prise_type_ef', 'prise_type_2', 'prise_type_combo_ccs', 'prise_type_chademo', 
           'prise_type_autre', 'paiement_acte', 'paiement_autre', 'reservation',  'accessibilite_pmr', 'restriction_gabarit', 
           'date_maj', 'last_modified']

In [187]:
data = irve.loc[:, champs]
data[['id_station_itinerance','id_pdc_itinerance']] = data[['id_station_itinerance','id_pdc_itinerance']].astype('string')
#data[['last_modified','date_maj']] = data[['last_modified','date_maj']].astype('datetime64')
data['non_concerne'] = data['id_station_itinerance'].str.contains('oncern') | data['id_pdc_itinerance'].str.contains('oncern')
non_concerne = data[data['non_concerne']].loc[:, champs].reset_index()
itinerance = data[data['non_concerne'] == False].loc[:, champs].reset_index()
print(len(non_concerne), len(itinerance))

itinerance_maj = itinerance.sort_values(by='date_maj').drop_duplicates('id_pdc_itinerance', keep='last').reset_index(drop=True)
print(len(itinerance_maj))
#idxs2 = Ilist.obj(data2)
#print(itinerance.dtypes)

607 50669
42368


In [197]:
itinerance['stat_maj'] = itinerance.id_station_itinerance + itinerance.date_maj
st_last = itinerance.sort_values(by='stat_maj').drop_duplicates('id_station_itinerance', keep='last')
itinerance['last'] = itinerance['stat_maj'].isin(st_last['stat_maj'])
itinerance_stat_maj = itinerance[itinerance['last']].copy()
itinerance_stat_maj = itinerance_stat_maj.drop(['last', 'stat_maj'], axis=1).reset_index(drop=True)
print(len(itinerance_stat_maj))
print(itinerance_stat_maj.dtypes)

38627
index                      int64
contact_operateur         object
nom_enseigne              object
coordonneesXY             object
adresse_station           object
id_station_itinerance     string
nom_station               object
implantation_station      object
nbre_pdc                   int64
condition_acces           object
horaires                  object
station_deux_roues        object
id_pdc_itinerance         string
puissance_nominale       float64
prise_type_ef             object
prise_type_2              object
prise_type_combo_ccs      object
prise_type_chademo        object
prise_type_autre          object
paiement_acte             object
paiement_autre            object
reservation               object
accessibilite_pmr         object
restriction_gabarit       object
date_maj                  object
last_modified             object
dtype: object


# 5 - Contrôle du schéma de données

Les contrôles précédents peuvent être réalisés à partir de la propriété 'relationship' définie dans le schéma de données.

## schéma de données
Le schéma de données restreint à la propriété 'relationship' et construit à partir du modèle de données est le suivants :

In [194]:
fields = [
 
 # relation unicité des pdl
 { "name": "index",
   "relationship" : { "parent" : "id_pdc_itinerance", "link" : "coupled" }},   
 # relations inter entités
 { "name": "contact_operateur",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nom_enseigne",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "coordonneesXY",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "id_station_itinerance",
   "relationship" : { "parent" : "id_pdc_itinerance",     "link" : "derived" }},
 # relations intra entité - station
 { "name": "nom_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "implantation_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nbre_pdc",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "condition_acces",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "horaires",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "station_deux_roues",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 # relations intra entité - localisation
 { "name": "adresse_station",
   "relationship" : { "parent" : "coordonneesXY",         "link" : "derived" }} ]

## Analyse et résultats
L'analyse fournit la liste des enregistrements incohérents pour chaque propriété.
Des champs booléens (valeur False à chaque enregistrement incohérent) pour chaque propriété sont ajoutés au jeu de données.

In [198]:
# génération du dictionnaire contenant la liste des enregistrements incohérent par propriété
data = itinerance_stat_maj #itinerance_maj
analyse = Analysis(data)

In [199]:
dic_res = analyse.check_relationship(fields)

# ajout d'un champ pour chaque propriété contrôlée (plus deux champs de synthèse)
data['total'] = True
data['somme'] = 0
for name, lis in dic_res.items():
    data[name] = True
    data.loc[lis, name] = False
    data['total'] = data['total'] & data[name]
    data['somme'] = data['somme'] - data[name] + 1
    print('{:<50} {:>5}'.format(name, len(data) - data[name].sum()))
maxi = max(data['somme'])
df_max = data[data.somme >= maxi]

# résultats
print("\nnombre d'enregistrements avec au moins une erreur : ", len(data) - data.total.sum())
print("nombre d'enregistrements avec ", maxi, " erreurs : ", len(df_max))
print("nombre d'enregistrements sans erreurs : ", data.total.sum())
# nb maxi d'erreurs
df_max.loc[:, relations]

index - id_pdc_itinerance                           2044
contact_operateur - id_station_itinerance              2
nom_enseigne - id_station_itinerance                 607
coordonneesXY - id_station_itinerance                888
id_station_itinerance - id_pdc_itinerance            447
nom_station - id_station_itinerance                  457
implantation_station - id_station_itinerance          89
nbre_pdc - id_station_itinerance                     616
condition_acces - id_station_itinerance                0
horaires - id_station_itinerance                     120
station_deux_roues - id_station_itinerance           325
adresse_station - coordonneesXY                      498

nombre d'enregistrements avec au moins une erreur :  3487
nombre d'enregistrements avec  5  erreurs :  2
nombre d'enregistrements sans erreurs :  35140


contact_operateur nom_enseigne            coordonneesXY  \
30592  contact@qovoltis.com     QOVOLTIS  [4.0309288, 49.2849379]   
31011  contact@qovoltis.com     QOVOLTIS    [4.0303445, 49.28518]   

                                         adresse_station  \
30592  ZONE JEAN XXIII, RUE DE LA FERME PIERQUIN, 1èr...   
31011  RUE DE LA FERME PIERQUIN, 1ère Classe 51450 BE...   

      id_station_itinerance                                     nom_station  \
30592         FRQOVP5100002  ZONE JEAN XXIII, RUE DE LA FERME PIERQUIN, 1èr   
31011         FRQOVP5100002   ZONE JEAN XXIII, RUE DE LA FERME PIERQUIN, 1è   

               implantation_station  nbre_pdc condition_acces  \
30592  Parking privé à usage public         7     Accès libre   
31011  Parking privé à usage public         7     Accès libre   

                horaires station_deux_roues id_pdc_itinerance    date_maj  \
30592  Mo-Fr 09:00-19:00                  0      FRQOVE182085  2022-08-11   
31011               24/7                  0      FRQOVE182085  2022-08-11   

                    last_modified  
30592  2023-03-03T17:36:05.518000  
31011  2023-03-03T17:35:52.123000

In [202]:
data.loc[~data['index - id_pdc_itinerance'], relations]

contact_operateur               nom_enseigne  \
3         contact@car2plug.fr          SUN DEVELOPPEMENT   
4         contact@car2plug.fr          SUN DEVELOPPEMENT   
5         contact@car2plug.fr               STATION CIEL   
363    info@rossinienergy.com              RossiniEnergy   
540    info@rossinienergy.com              RossiniEnergy   
...                       ...                        ...   
38610  contact@ubitricity.com  Ubitricity France Network   
38611  contact@ubitricity.com  Ubitricity France Network   
38612  contact@ubitricity.com  Ubitricity France Network   
38613  contact@ubitricity.com  Ubitricity France Network   
38614  contact@ubitricity.com  Ubitricity France Network   

                                 coordonneesXY  \
3       [2.316255993913532, 48.87284863189276]   
4       [2.316255993913532, 48.87284863189276]   
5      [4.925962686538696, 45.746076812151216]   
363                      [2.027372, 50.367401]   
540                      [2.816443, 49.417865]   
...                                        ...   
38610                    [0.163932, 49.513496]   
38611                    [0.172373, 49.499466]   
38612                    [0.191899, 49.506711]   
38613                    [0.193942, 49.544211]   
38614                    [0.193942, 49.544211]   

                                     adresse_station id_station_itinerance  \
3                   360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
4                   360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
5                      \t378 RTE DE GENAS 69500 BRON           FRC2P000301   
363                     D136, 62140 Marconne, France              FRROSE15   
540    Rue des Bateliers60280 Margny-lès-Compiègne\n             FRROSE175   
...                                              ...                   ...   
38610                            Parc relais Schuman        FRUBIE10017845   
38611                        Stade Oc�ane - Le Havre        FRUBIE10084198   
38612    Avenue Pr�sident Ren� Coty (Gare Beaulieu)         FRUBIE10091696   
38613                              Place Abb� Pierre        FRUBIE10056568   
38614                              Place Abb� Pierre        FRUBIE10058428   

                        nom_station          implantation_station  nbre_pdc  \
3                SUN DEVELOPPEMENT   Parking privé à usage public         2   
4                SUN DEVELOPPEMENT   Parking privé à usage public         2   
5                      STATION CIEL  Parking privé à usage public         1   
363                   renaulthesdin  Parking privé à usage public         1   
540                      Taim_Hotel  Parking privé à usage public         6   
...                             ...                           ...       ...   
38610  Caucriauville Lut - Le Havre                Parking public         1   
38611            Vallee Bereult Sud                Parking public         1   
38612                      Beaulieu                        Voirie         1   
38613  Centre ville - Montivilliers                Parking public         1   
38614  Centre ville - Montivilliers                Parking public         1   

      condition_acces horaires station_deux_roues id_pdc_itinerance  \
3         Accès libre     24/7              false      FRC2P0005012   
4         Accès libre     24/7              false      FRC2P0005011   
5         Accès libre     24/7              false      FRC2P0003011   
363       Accès libre     24/7              FALSE         FRROSE151   
540       Accès libre     24/7              FALSE         FRROSE175   
...               ...      ...                ...               ...   
38610     Accès libre     24/7              FALSE    FRUBIE10017845   
38611     Accès libre     24/7              FALSE    FRUBIE10084198   
38612     Accès libre     24/7              FALSE    FRUBIE10091696   
38613     Accès libre     24/7              FALSE    FRUBIE10056568   
38614     Accès libre  

In [204]:
data.loc[data.id_station_itinerance == 'FRC2P000501', relations]

contact_operateur       nom_enseigne  \
3      contact@car2plug.fr  SUN DEVELOPPEMENT   
4      contact@car2plug.fr  SUN DEVELOPPEMENT   
17241  contact@car2plug.fr  SUN DEVELOPPEMENT   
17242  contact@car2plug.fr  SUN DEVELOPPEMENT   
17328  contact@car2plug.fr  SUN DEVELOPPEMENT   
17333  contact@car2plug.fr  SUN DEVELOPPEMENT   
17339  contact@car2plug.fr  SUN DEVELOPPEMENT   
17340  contact@car2plug.fr  SUN DEVELOPPEMENT   

                                coordonneesXY  \
3      [2.316255993913532, 48.87284863189276]   
4      [2.316255993913532, 48.87284863189276]   
17241  [1.683051586151123, 48.59249496459961]   
17242  [1.683051586151123, 48.59249496459961]   
17328  [2.316255993913532, 48.87284863189276]   
17333  [2.316255993913532, 48.87284863189276]   
17339  [1.683051586151123, 48.59249496459961]   
17340  [1.683051586151123, 48.59249496459961]   

                        adresse_station id_station_itinerance  \
3      360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
4      360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17241  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17242  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17328  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17333  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17339  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   
17340  360 AV DE L EUROPE 28230 EPERNON           FRC2P000501   

              nom_station          implantation_station  nbre_pdc  \
3      SUN DEVELOPPEMENT   Parking privé à usage public         2   
4      SUN DEVELOPPEMENT   Parking privé à usage public         2   
17241   SUN DEVELOPPEMENT  Parking privé à usage public         2   
17242   SUN DEVELOPPEMENT  Parking privé à usage public         2   
17328   SUN DEVELOPPEMENT  Parking privé à usage public         2   
17333   SUN DEVELOPPEMENT  Parking privé à usage public         2   
17339   SUN DEVELOPPEMENT  Parking privé à usage public         2   
17340   SUN DEVELOPPEMENT  Parking privé à usage public         2   

      condition_acces horaires station_deux_roues id_pdc_itinerance  \
3         Accès libre     24/7              false      FRC2P0005012   
4         Accès libre     24/7              false      FRC2P0005011   
17241     Accès libre     24/7              false      FRC2P0005012   
17242     Accès libre     24/7              false      FRC2P0005011   
17328     Accès libre     24/7              false      FRC2P0005011   
17333     Accès libre     24/7              false      FRC2P0005012   
17339     Accès libre     24/7              false      FRC2P0005012   
17340     Accès libre     24/7              false      FRC2P0005011   

         date_maj               last_modified  
3      2021-07-25  2023-03-03T17:43:52.865000  
4      2021-07-25  2023-03-03T17:43:52.865000  
17241  2021-07-25  2023-03-03T17:38:22.684000  
17242  2021-07-25  2023-03-03T17:38:22.684000  
17328  2021-07-25  2023-03-03T17:38:19.237000  
17333  2021-07-25  2023-03-03T17:38:19.237000  
17339  2021-07-25  2023-03-03T17:38:16.571000  
17340  2021-07-25  2023-03-03T17:38:16.571000

index - id_pdc_itinerance                          16123
contact_operateur - id_station_itinerance          10719
nom_enseigne - id_station_itinerance                7514
coordonneesXY - id_station_itinerance              11825
id_station_itinerance - id_pdc_itinerance            578
nom_station - id_station_itinerance                 1865
implantation_station - id_station_itinerance        1245
nbre_pdc - id_station_itinerance                    1458
condition_acces - id_station_itinerance               35
horaires - id_station_itinerance                    9869
station_deux_roues - id_station_itinerance         10968
adresse_station - coordonneesXY                     1360

nombre d'enregistrements avec au moins une erreur :  18116
nombre d'enregistrements avec  10  erreurs :  2
nombre d'enregistrements sans erreurs :  32553

In [203]:
itinerance_maj.loc[itinerance_maj.id_station_itinerance == 'FRH14P59309001', relations]

contact_operateur          nom_enseigne  \
6041   support@passpasselectrique.fr  Pass Pass Electrique   
13487         support@alizecharge.fr  pass pass électrique   
13488         support@alizecharge.fr  pass pass électrique   

               coordonneesXY                    adresse_station  \
6041   [2.579956, 50.978392]  4 rue de Cassel 59122 Hondschoote   
13487  [2.579956, 50.978392]  4 Rue De Cassel 59122 HONDSCHOOTE   
13488  [2.579956, 50.978392]  4 Rue De Cassel 59122 HONDSCHOOTE   

      id_station_itinerance                   nom_station  \
6041         FRH14P59309001  Hondschoote - Rue de Cassel    
13487        FRH14P59309001   HONDSCHOOTE - Rue De Cassel   
13488        FRH14P59309001   HONDSCHOOTE - Rue De Cassel   

      implantation_station  nbre_pdc condition_acces           horaires  \
6041        Parking public         1     Accès libre               24/7   
13487               Voirie         2     Accès libre  Mo-Su 00:00-23:59   
13488               Voirie         2     Accès libre  Mo-Su 00:00-23:59   

      station_deux_roues id_pdc_itinerance    date_maj  \
6041               false    FRH14P59309001  2022-08-16   
13487              FALSE   FRH14E593090011  2022-12-15   
13488              FALSE   FRH14E593090012  2022-12-15   

                    last_modified  
6041   2023-03-03T17:38:38.448000  
13487  2023-03-03T17:39:32.426000  
13488  2023-03-03T17:39:32.426000

In [174]:
FRA68P68021001 = itinerance[itinerance.id_station_itinerance == 'FRA68P68021001']
FRA68P68021001.loc[:, relations]

contact_operateur                  nom_enseigne  \
1685          hostellerie.des.ducs@wanadoo.fr                Charge parking   
1686                   sas.masseria@gmail.com                Charge banarge   
9070            contact@societe-operateur.com        Réseau de recharge ABC   
42022                             rcgc@edf.fr                           EDF   
42068                      contact@ekoplug.fr                         aucun   
42276               ssznek@themis-experts.com                           edf   
42279                         g.arpino@ocr.fr  OCR MAINTENANCE ELECTRONIQUE   
42302                      contact@energem.fr                       ENERGEM   
42303  edfentreprises-ile-de-france-31@edf.fr                           EDF   
42304          electricite@alterna-energie.fr               alterna energie   

                                   coordonneesXY  \
1685   [0.18225850129804153, 44.677126551459914]   
1686                     [0.1984829, 44.6656545]   
9070                       [7.487105, 48.345345]   
42022                               [2.02,48.53]   
42068                               [4.61,46.34]   
42276                               [2.24,48.94]   
42279                               [2.23,48.69]   
42302                               [2.63,48.98]   
42303                               [2.63,48.89]   
42304                               [2.25,48.70]   

                             adresse_station id_station_itinerance  \
1685                        bd jean brisseau        FRA68P68021001   
1686                        za banarge duras        FRA68P68021001   
9070         1 avenue de la Paix 75001 Paris        FRA68P68021001   
42022             15 rue de l'Ermitage 91410        FRA68P68021001   
42068  330 route de Chagny 71520 SAINT-POINT        FRA68P68021001   
42276                  89 RUE HENRI BARBUSSE        FRA68P68021001   
42279                3 AVENUE DE OUESSANT 91        FRA68P68021001   
42302        36 rue Denis Papin 77290 Mitry         FRA68P68021001   
42303                        RUE CURIE 77177        FRA68P68021001   
42304                 24 Rue du Baron de Niv        FRA68P68021001   

                  nom_station                  implantation_station  nbre_pdc  \
1685                  parking  Parking privé réservé à la clientèle         1   
1686                  Banarge          Parking privé à usage public         1   
9070                   Picpus                                Voirie        10   
42022                 Dourdan          Parking privé à usage public         4   
42068  gite danse des etoiles  Parking privé réservé à la clientèle         3   
42276              ARGENTEUIL  Parking privé réservé à la clientèle         3   
42279     Villebon sur yvette  Parking privé réservé à la clientèle         2   
42302              Mitry Mory  Parking privé réservé à la clientèle         4   
42303    BROU SUR CHANTEREINE          Parking privé à usage public         2   
42304     Villebon sur yvette          Parking privé à usage public         4   

      condition_acces                                           horaires  \
1685      Accès libre                                               24/7   
1686      Accès libre                                               24/7   
9070      Accès libre  Mo-Fr 08:00-12:00,Mo-Fr 14:00-18:00,Th 08:00-1...   
42022   Accès réservé               Mo-Fr 08:00-12:00, Mo-Fr 14:00-18:00   
42068   Accès réservé                                               24/7   
42276   Accès réservé  Mo-Fr 08:00-12:00,Mo-Fr 14:00-18:00, Th 08:00-...   
42279   Accès réservé               Mo-Fr 08:00-12:00, Mo-Fr 14:00-18:00   
42302   Accès réservé               Mo-Fr 08:00-12:00, Mo-Fr 14:00-18:00   
42303     Accès libre               Mo-Fr 08:00-12:00, Mo-Fr 14:00-18:00   
42304   Accès réservé               Mo-Fr 08:00-12:00, Mo-Fr 14:00-18:00   

      station_deux_roues id_pdc_itinerance    date_maj  \
1685               false   

In [132]:
coupled = ~itinerance['coordonneesXY - id_station_itinerance'] & ~itinerance['id_station_itinerance - coordonneesXY']
print(coupled.sum(), len(coupled) - itinerance['coordonneesXY - id_station_itinerance'].sum(), 
      len(coupled) - itinerance['id_station_itinerance - coordonneesXY'].sum())

309 11825 8414


----
## initialisation de l'outil d'analyse
L'outil d'analyse ([voir présentation](https://github.com/loco-philippe/Environmental-Sensing/blob/main/documentation/FR_tabular_structure.ipynb)) est chargé à partir des données Pandas :
On dénombre 211 370 valeurs différentes sur un total de 51 276 x 50 valeurs ("taux d'unicité" de 8.2%)

In [71]:
idxs = Ilist.obj(itinerance)
res['lignes_data'] = len(idxs)
res['champs_data'] = len(idxs.idxlen)
res['unique_data'] = sum(idxs.idxlen)
print('nombre de lignes              : ', res['lignes_data'])
print('nombre de champs              : ', res['champs_data'])
print('nombre de valeurs différentes : ', res['unique_data'])

nombre de lignes              :  50669
nombre de champs              :  42
nombre de valeurs différentes :  213643


_________________
# 3 - Analyse des relations entre champs
## Arborescence des champs
- l'examen de "l'arbre de dérivation" indique un faible niveau d'interdépendance entre champs (seuls quelques-uns ne sont pas liés à la racine). 
- Par exemple les champs consolidated_longitude(44) et consolidated_latitude(45) sont bien dérivés du champ coordonneesXY(13)


In [6]:
print(idxs.tree())
# valeur entre parenthèses : nombre de valeurs distinctes

-1: root-derived (51276)
   0 : nom_amenageur (2474)
   1 : siren_amenageur (1182)
   2 : contact_amenageur (548)
   3 : nom_operateur (237)
   4 : contact_operateur (274)
   5 : telephone_operateur (357)
   6 : nom_enseigne (1493)
   7 : id_station_itinerance (18544)
   8 : id_station_local (10731)
   9 : nom_station (15463)
   10: implantation_station (5)
   11: adresse_station (16496)
   12: code_insee_commune (5242)
   13: coordonneesXY (17241)
      44: consolidated_longitude (17069)
      45: consolidated_latitude (17018)
   14: nbre_pdc (50)
   15: id_pdc_itinerance (42457)
   16: id_pdc_local (27186)
   17: puissance_nominale (87)
   18: prise_type_ef (8)
   19: prise_type_2 (8)
   20: prise_type_combo_ccs (8)
   21: prise_type_chademo (7)
   22: prise_type_autre (7)
   23: gratuit (8)
   24: paiement_acte (8)
   25: paiement_cb (7)
   26: paiement_autre (8)
   27: tarification (212)
   28: condition_acces (2)
   29: reservation (7)
   30: horaires (516)
   31: accessibilite_pm

# 4 - Contrôle des relations
Le contrôle consiste à valider les relations inter entités ainsi que les relations intra entités. Pour le modèle identifié précédemment, les contrôles sont les suivants :
- contrôle d'unicité des lignes (une ligne est associée à un point de charge)
- contrôle inter entités (pdc-station, station-opérateur, station-localisation, station-enseigne)
- contrôle intra entités (station, localisation)

## unicité des lignes
- vérification de la relation 1-1 entre le champ 'root' (liste des lignes) et le champ 'id_pdc_itinerance' (cf analyse au chapitre 5)
- le nombre de lignes en écart est de 16642 (avec historique) et 1068 (hors historique des modifications)

In [7]:
champ = idxs.nindex
root = Iindex.ext(list(range(len(idxs))))

ecart_root = champ('id_pdc_itinerance').coupling(root, reindex=True)

print('lignes en écart : ',len(ecart_root))

lignes en écart :  16642


## relations inter-entités
L'analyse est effectuée sur la base des données hors historique (data2).    
Les relation 1-n avec les points de charge n'ont pas besoin d'être controlées (unicité des points de charge).
   
Nota :
- *si on transforme la relation 1-n entre station et localisation en une relation 1-1, les écarts sont beaucoups plus importants (plus de 7 000 lignes), ceci indique qu'à une même localisation on trouve plusieurs stations (ex. identifiants de points de charges utilisés comme identifiants de stations)*

In [8]:
data2 = data.sort_values(by='last_modified').drop_duplicates('id_pdc_itinerance', keep='last').reset_index(drop=True)
idxs2 = Ilist.obj(data2)
champ = idxs2.nindex

In [9]:
ecart_oper = champ('id_station_itinerance').coupling(champ('contact_operateur'), reindex=True)
ecart_XY = champ('id_station_itinerance').coupling(champ('coordonneesXY'), reindex=True)
ecart_ens = champ('id_station_itinerance').coupling(champ('nom_enseigne'), reindex=True)

print('écart relation id_station - contact_operateur : ', len(ecart_oper), ' lignes')
print('écart relation id_station - coordonneesXY     : ', len(ecart_XY), ' lignes')
print('écart relation id_station - nom_enseigne      : ', len(ecart_ens), ' lignes')

écart relation id_station - contact_operateur :  285  lignes
écart relation id_station - coordonneesXY     :  705  lignes
écart relation id_station - nom_enseigne      :  547  lignes


## relations intra entités
On ne considère que les champs obligatoires qui doivent avoir une relation 1-n avec le champ 'id_station_itinerance' pour les stations, 'coordonneesXY' pour les localisations.

In [10]:
ecart_nom = champ('id_station_itinerance').coupling(champ('nom_station'), reindex=True)
ecart_imp = champ('id_station_itinerance').coupling(champ('implantation_station'), reindex=True)
ecart_nbr = champ('id_station_itinerance').coupling(champ('nbre_pdc'), reindex=True)
ecart_acc = champ('id_station_itinerance').coupling(champ('condition_acces'), reindex=True)
ecart_hor = champ('id_station_itinerance').coupling(champ('horaires'), reindex=True)
ecart_rou = champ('id_station_itinerance').coupling(champ('station_deux_roues'), reindex=True)

ecart_adr = champ('coordonneesXY').coupling(champ('adresse_station'), reindex=True)


print('écart relation id_station - nom_station          : ',len(ecart_nom), ' lignes')
print('écart relation id_station - implantation_station : ',len(ecart_imp), ' lignes')
print('écart relation id_station - nbre_pdc             : ',len(ecart_nbr), ' lignes')
print('écart relation id_station - condition_acces      : ',len(ecart_acc), '   lignes')
print('écart relation id_station - horaires             : ',len(ecart_hor), ' lignes')
print('écart relation id_station - station_deux_roues   : ',len(ecart_rou), ' lignes', '\n')

print('écart relation coordonneesXY - adresse_station   : ',len(ecart_adr), ' lignes')

écart relation id_station - nom_station          :  288  lignes
écart relation id_station - implantation_station :  217  lignes
écart relation id_station - nbre_pdc             :  597  lignes
écart relation id_station - condition_acces      :  5    lignes
écart relation id_station - horaires             :  263  lignes
écart relation id_station - station_deux_roues   :  616  lignes 

écart relation coordonneesXY - adresse_station   :  604  lignes


In [22]:
# génération du dictionnaire contenant la liste des enregistrements incohérent par propriété
analyse = Analysis(data2)
dic_res = analyse.check_relationship(fields)

# ajout d'un champ pour chaque propriété contrôlée (plus deux champs de synthèse)
data2['total'] = True
data2['somme'] = 0
for name, lis in dic_res.items():
    data2[name] = True
    data2.loc[lis, name] = False
    data2['total'] = data2['total'] & data2[name]
    data2['somme'] = data2['somme'] - data2[name] + 1
    print('{:<50} {:>5}'.format(name, len(data2) - data2[name].sum()))
maxi = max(data2['somme'])
df_max = data2[data2.somme == maxi]

# résultats
print("nombre d'enregistrements avec au moins une erreur : ", len(data2) - data2.total.sum())
print("nombre d'enregistrements avec ", maxi, " erreurs : ", len(df_max))
df_max.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY')]

contact_operateur - id_station_itinerance            285
nom_enseigne - id_station_itinerance                 547
coordonneesXY - id_station_itinerance                705
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                  288
implantation_station - id_station_itinerance         217
nbre_pdc - id_station_itinerance                     597
condition_acces - id_station_itinerance                5
horaires - id_station_itinerance                     263
station_deux_roues - id_station_itinerance           616
adresse_station - coordonneesXY                      604
nombre d'enregistrements avec au moins une erreur :  2218
nombre d'enregistrements avec  9  erreurs :  14


id_pdc_itinerance id_station_itinerance                     nom_station  \
7923        FRE10E40150          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7924        FRE10E40151          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7925        FRE10E40149          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7926        FRE10E40152          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7927        FRE10E40154          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7928        FRE10E40155          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7929        FRE10E40148          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
7930        FRE10E40153          Non concerné  HERVE THERMIQUE JOUE LES TOURS   
41082    FRLDLE00001763          Non concerné          LIDL ST AUBIN SUR SCIE   
41083    FRLDLE00001764          Non concerné          LIDL ST AUBIN SUR SCIE   
41099    FRLDLE00001743          Non concerné                     LIDL YVETOT   
41102    FRLDLE00001744          Non concerné                     LIDL YVETOT   
41107    FRLDLE00001779          Non concerné                   LIDL BARENTIN   
41112    FRLDLE00001780          Non concerné                   LIDL BARENTIN   

                coordonneesXY  
7923    [0.654826, 47.353958]  
7924    [0.654826, 47.353958]  
7925    [0.654826, 47.353958]  
7926    [0.654826, 47.353958]  
7927    [0.654826, 47.353958]  
7928    [0.654826, 47.353958]  
7929    [0.654826, 47.353958]  
7930    [0.654826, 47.353958]  
41082  [0.9605582,49.5355585]  
41083  [0.9605582,49.5355585]  
41099  [0.9605582,49.5355585]  
41102  [0.9605582,49.5355585]  
41107  [0.9605582,49.5355585]  
41112  [0.9605582,49.5355585]

In [23]:
df_non_concerne = data2[data2.id_station_itinerance == 'Non concerné']
df_concerne = data2[data2.id_station_itinerance != 'Non concerné']
print("nombre d'enregistrements avec station 'non concerné' : ", len(df_non_concerne))

df_concerne_max_1 = df_concerne[df_concerne.somme == maxi-1]
print("nombre d'enregistrements avec ", maxi-1, " erreurs : ", len(df_concerne_max_1))
df_concerne_max_1.loc[:, ('id_pdc_itinerance', 'id_station_itinerance')]

nombre d'enregistrements avec station 'non concerné' :  89
nombre d'enregistrements avec  8  erreurs :  6


id_pdc_itinerance id_station_itinerance
21818    FRH14P59663001        FRH14P59663001
21831    FRH14P59309001        FRH14P59309001
35098   FRH14E593090012        FRH14P59309001
35099   FRH14E593090011        FRH14P59309001
35100   FRH14E596630012        FRH14P59663001
35101   FRH14E596630011        FRH14P59663001

In [24]:
FRH14P59309001 = data2[data2.id_station_itinerance == 'FRH14P59309001']
FRH14P59309001.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'adresse_station', 
                          'contact_operateur', 'nom_enseigne', 'implantation_station', 'nbre_pdc', 'horaires', 
                          'condition_acces', 'station_deux_roues')]

id_pdc_itinerance id_station_itinerance                   nom_station  \
21831    FRH14P59309001        FRH14P59309001  Hondschoote - Rue de Cassel    
35098   FRH14E593090012        FRH14P59309001   HONDSCHOOTE - Rue De Cassel   
35099   FRH14E593090011        FRH14P59309001   HONDSCHOOTE - Rue De Cassel   

               coordonneesXY                    adresse_station  \
21831  [2.579956, 50.978392]  4 rue de Cassel 59122 Hondschoote   
35098  [2.579956, 50.978392]  4 Rue De Cassel 59122 HONDSCHOOTE   
35099  [2.579956, 50.978392]  4 Rue De Cassel 59122 HONDSCHOOTE   

                   contact_operateur          nom_enseigne  \
21831  support@passpasselectrique.fr  Pass Pass Electrique   
35098         support@alizecharge.fr  pass pass électrique   
35099         support@alizecharge.fr  pass pass électrique   

      implantation_station  nbre_pdc           horaires condition_acces  \
21831       Parking public         1               24/7     Accès libre   
35098               Voirie         2  Mo-Su 00:00-23:59     Accès libre   
35099               Voirie         2  Mo-Su 00:00-23:59     Accès libre   

      station_deux_roues  
21831              false  
35098              FALSE  
35099              FALSE

In [25]:
FRH14P59663001 = data2[data2.id_station_itinerance == 'FRH14P59663001']
FRH14P59663001.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'adresse_station', 
                          'contact_operateur', 'nom_enseigne', 'implantation_station', 'nbre_pdc', 'horaires', 
                          'condition_acces', 'station_deux_roues')]

id_pdc_itinerance id_station_itinerance  \
21818    FRH14P59663001        FRH14P59663001   
35100   FRH14E596630012        FRH14P59663001   
35101   FRH14E596630011        FRH14P59663001   

                                 nom_station          coordonneesXY  \
21818              Wormhout -Place de Gaulle   [2.46799, 50.880845]   
35100  WORMHOUT - Place Du General De Gaulle  [2.467585, 50.883458]   
35101  WORMHOUT - Place Du General De Gaulle  [2.467585, 50.883458]   

                                    adresse_station  \
21818             92 Place De Gaulle 59470 Wormhout   
35100  92 Place Du General De Gaulle 59470 WORMHOUT   
35101  92 Place Du General De Gaulle 59470 WORMHOUT   

                   contact_operateur          nom_enseigne  \
21818  support@passpasselectrique.fr  Pass Pass Electrique   
35100         support@alizecharge.fr  pass pass électrique   
35101         support@alizecharge.fr  pass pass électrique   

      implantation_station  nbre_pdc           horaires condition_acces  \
21818       Parking public         2               24/7     Accès libre   
35100               Voirie         2  Mo-Su 00:00-23:59     Accès libre   
35101               Voirie         2  Mo-Su 00:00-23:59     Accès libre   

      station_deux_roues  
21818              false  
35100              FALSE  
35101              FALSE

# 6 - Exemples d'analyse de relations
## Identification du champ 'root'
Le champ 'root' est l'identifiant d'une ligne du tableau. Dans le cas présent, aucun champ ne remplit ce rôle. Le champ le plus proche est le champ 'id_pdc_itinerance' avec 42 457 valeurs pour 51 276 lignes.

In [44]:
champ = idxs.nindex
root = Iindex.ext(list(range(len(idxs))))

ecart = champ('id_pdc_itinerance').coupling(root)
print('lignes en écart : ',len(ecart))

c = Counter(champ('id_pdc_itinerance').codec).most_common(50)
champ('id_pdc_itinerance').reindex()
print('les 50 pdc avec le plus de lignes : \n', c, '\n')

lignes en écart :  16642
les 50 pdc avec le plus de lignes : 
 [('Non concerné', 519), ('FR55CP84000', 50), ('FR55CP92140', 29), ('FR55CP78140', 28), ('FR55CP83310', 23), ('FR55CP94340', 18), ('FR55CP92260', 16), ('FR55CP78000', 16), ('FR55CP92380', 12), ('FR55CP92130', 12), ('FR55CP83120', 10), ('FRA68P680210015', 10), ('FR55CP83990', 10), ('FR55CP83350', 9), ('FR55CPBP73800', 8), ('FR55CPBP73190', 8), ('FR55CP92280', 6), ('FR55CP30133', 6), ('FR55CPBP514', 6), ('FR55CP33240', 6), ('FR55CPBP698', 6), ('FR55CP92320', 6), ('FRLMSE117226622', 5), ('FRLMSE117201901', 5), ('FRLMSE117226621', 5), ('FRLMSE117228771', 5), ('FRLMSE117201891', 5), ('FRLMSP117226962', 5), ('FRLMSE117228772', 5), ('FRLMSP117226961', 5), ('FRFASE3302702', 4), ('FRS33EMB3315611', 4), ('FRS33EMB3315811', 4), ('FRFASE3302508', 4), ('FRFASE3302405', 4), ('FRFASE3302204', 4), ('FRFASE3302201', 4), ('FRFASE3302504', 4), ('FRFASE3302803', 4), ('FRFASE3302506', 4), ('FRFASE3302802', 4), ('FRS33EMB3315211', 4), ('FRFASE330

In [45]:
FRFASE3302502 = data[data.id_pdc_itinerance == 'FRFASE3302502']
FRFASE3302502.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'nbre_pdc', 'last_modified')]

id_pdc_itinerance id_station_itinerance             nom_station  \
1991      FRFASE3302502           FRFASE33025  Aire de Saint Ambreuil   
2122      FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   
2175      FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   
42539     FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   

                                coordonneesXY  nbre_pdc  \
1991                     [4.844517, 46.69318]         8   
2122   [4.844761236754013, 46.69092869473468]         8   
2175   [4.844761236754013, 46.69092869473468]         8   
42539  [4.844761236754013, 46.69092869473468]         8   

                    last_modified  
1991   2023-03-03T17:40:02.071000  
2122   2023-03-03T17:39:56.927000  
2175   2023-03-03T17:39:56.627000  
42539  2023-03-03T17:36:39.503000

In [46]:
FRFASE33025 = data[data.id_station_itinerance == 'FRFASE33025']
FRFASE33025.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'nbre_pdc', 'last_modified')]

id_pdc_itinerance id_station_itinerance             nom_station  \
1957     FRFASE3302508           FRFASE33025  Aire de Saint Ambreuil   
1958     FRFASE3302503           FRFASE33025  Aire de Saint Ambreuil   
1959     FRFASE3302507           FRFASE33025  Aire de Saint Ambreuil   
1960     FRFASE3302506           FRFASE33025  Aire de Saint Ambreuil   
1961     FRFASE3302505           FRFASE33025  Aire de Saint Ambreuil   
1962     FRFASE3302504           FRFASE33025  Aire de Saint Ambreuil   
1968     FRFASE3302501           FRFASE33025  Aire de Saint Ambreuil   
1991     FRFASE3302502           FRFASE33025  Aire de Saint Ambreuil   

             coordonneesXY  nbre_pdc               last_modified  
1957  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1958  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1959  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1960  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1961  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1962  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1968  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000  
1991  [4.844517, 46.69318]         8  2023-03-03T17:40:02.071000

In [47]:
FRFASP33025 = data[data.id_station_itinerance == 'FRFASP33025']
FRFASP33025.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'nbre_pdc', 'last_modified')]

id_pdc_itinerance id_station_itinerance             nom_station  \
2122      FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   
2131      FRFASE3302501           FRFASP33025  Aire de Saint Ambreuil   
2133      FRFASE3302503           FRFASP33025  Aire de Saint Ambreuil   
2134      FRFASE3302505           FRFASP33025  Aire de Saint Ambreuil   
2135      FRFASE3302506           FRFASP33025  Aire de Saint Ambreuil   
2136      FRFASE3302507           FRFASP33025  Aire de Saint Ambreuil   
2137      FRFASE3302508           FRFASP33025  Aire de Saint Ambreuil   
2141      FRFASE3302504           FRFASP33025  Aire de Saint Ambreuil   
2164      FRFASE3302503           FRFASP33025  Aire de Saint Ambreuil   
2166      FRFASE3302501           FRFASP33025  Aire de Saint Ambreuil   
2175      FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   
2177      FRFASE3302504           FRFASP33025  Aire de Saint Ambreuil   
2178      FRFASE3302505           FRFASP33025  Aire de Saint Ambreuil   
2179      FRFASE3302506           FRFASP33025  Aire de Saint Ambreuil   
2180      FRFASE3302507           FRFASP33025  Aire de Saint Ambreuil   
2181      FRFASE3302508           FRFASP33025  Aire de Saint Ambreuil   
42537     FRFASE3302504           FRFASP33025  Aire de Saint Ambreuil   
42538     FRFASE3302501           FRFASP33025  Aire de Saint Ambreuil   
42539     FRFASE3302502           FRFASP33025  Aire de Saint Ambreuil   
42540     FRFASE3302503           FRFASP33025  Aire de Saint Ambreuil   
42541     FRFASE3302506           FRFASP33025  Aire de Saint Ambreuil   
42545     FRFASE3302505           FRFASP33025  Aire de Saint Ambreuil   
42551     FRFASE3302508           FRFASP33025  Aire de Saint Ambreuil   
42567     FRFASE3302507           FRFASP33025  Aire de Saint Ambreuil   

                                coordonneesXY  nbre_pdc  \
2122   [4.844761236754013, 46.69092869473468]         8   
2131   [4.844761236754013, 46.69092869473468]         8   
2133   [4.844761236754013, 46.69092869473468]         8   
2134   [4.844761236754013, 46.69092869473468]         8   
2135   [4.844761236754013, 46.69092869473468]         8   
2136   [4.844761236754013, 46.69092869473468]         8   
2137   [4.844761236754013, 46.69092869473468]         8   
2141   [4.844761236754013, 46.69092869473468]         8   
2164   [4.844761236754013, 46.69092869473468]         8   
2166   [4.844761236754013, 46.69092869473468]         8   
2175   [4.844761236754013, 46.69092869473468]         8   
2177   [4.844761236754013, 46.69092869473468]         8   
2178   [4.844761236754013, 46.69092869473468]         8   
2179   [4.844761236754013, 46.69092869473468]         8   
2180   [4.844761236754013, 46.69092869473468]         8   
2181   [4.844761236754013, 46.69092869473468]         8   
42537  [4.844761236754013, 46.69092869473468]         8   
42538  [4.844761236754013, 46.69092869473468]         8   
42539  [4.844761236754013, 46.69092869473468]         8   
42540  [4.844761236754013, 46.69092869473468]         8   
42541  [4.844761236754013, 46.69092869473468]         8   
42545  [4.844761236754013, 46.69092869473468]         8   
42551  [4.844761236754013, 46.69092869473468]         8   
42567  [4.844761236754013, 46.69092869473468]         8   

                    last_modified  
2122   2023-03-03T17:39:56.927000  
2131   2023-03-03T17:39:56.927000  
2133   2023-03-03T17:39:56.927000  
2134   2023-03-03T17:39:56.927000  
2135   2023-03-03T17:39:56.927000  
2136   2023-03-03T17:39:56.927000  
2137   2023-03-03T17:39:56.927000  
2141   2023-03-03T17:39:56.927000  
2164   2023-03-03T17:39:56.627000  
2166   2023-03-03T17:39:56.627000  
2175   2023-03-03T17:39:56.627000  
2177   2023-03-03T17:39:56.627000  
2178   2023-03-03T17:39:56.627000  
2179   2023-03-03T17:39:56.627000  
2180   2023-03-03T17:39:56.627000  
2181   2023-03-03T17:39:56.627000  
42537  2023-03-03T17:36:39.503000  
42538  2023-03-03T17:36:39.503000  
42539  2023-03-03T17

In [42]:
FR55CP92140 = data[data.id_pdc_itinerance == 'FR55CP92140']
FR55CP92140.loc[:, ('id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'nbre_pdc', 'last_modified')]

id_pdc_itinerance id_station_itinerance  \
2388       FR55CP92140           FR55CP92140   
2407       FR55CP92140           FR55CP92140   
2408       FR55CP92140           FR55CP92140   
2409       FR55CP92140           FR55CP92140   
2410       FR55CP92140           FR55CP92140   
2411       FR55CP92140           FR55CP92140   
2419       FR55CP92140           FR55CP92140   
2420       FR55CP92140           FR55CP92140   
2421       FR55CP92140           FR55CP92140   
2423       FR55CP92140           FR55CP92140   
2424       FR55CP92140           FR55CP92140   
2425       FR55CP92140           FR55CP92140   
2426       FR55CP92140           FR55CP92140   
2427       FR55CP92140           FR55CP92140   
2428       FR55CP92140           FR55CP92140   
2429       FR55CP92140           FR55CP92140   
2430       FR55CP92140           FR55CP92140   
2469       FR55CP92140           FR55CP92140   
2499       FR55CP92140           FR55CP92140   
2500       FR55CP92140           FR55CP92140   
2514       FR55CP92140           FR55CP92140   
2516       FR55CP92140           FR55CP92140   
2517       FR55CP92140           FR55CP92140   
2518       FR55CP92140           FR55CP92140   
2519       FR55CP92140           FR55CP92140   
2574       FR55CP92140           FR55CP92140   
2587       FR55CP92140           FR55CP92140   
2588       FR55CP92140           FR55CP92140   
2609       FR55CP92140           FR55CP92140   

                                       nom_station          coordonneesXY  \
2388                            POMPIDOU - CLAMART   [2.23637, 48.784747]   
2407  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2408                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2409                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2410                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2411                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2419  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2420  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2421  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2423  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2424  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART  [2.255335, 48.786899]   
2425                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2426                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2427                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2428                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2429                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2430                       PLACE FERRARI - CLAMART  [2.262841, 48.798985]   
2469                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2499                            POMPIDOU - CLAMART   [2.23637, 48.784747]   
2500                         VICTOR HUGO - CLAMART  [2.267296, 48.804229]   
2514                  AVENUE JEAN JAURES - CLAMART  [2.267834, 48.808019]   
2516                  AVENUE JEAN JAURES - CLAMART  [2.267834, 48.808019]   
2517                  AVENUE JEAN JAURES - CLAMART  [2.267834, 48.808019]   
2518                  AVENUE JEAN JAURES - CLAMART  [2.267834, 48.808019]   
2519                  AVENUE JEAN JAURES - CLAMART  [2.267834, 48.808019]   
2574                            POMPIDOU - CLAMART   [2.23637, 48.784747]   
2587                            POMPIDOU - CLAMART   [2.23637, 48.784747]   
2588                            POMPIDOU - CLAMART   [2.23637, 48.784747]   
2609                            POMPIDOU - CLAMART   [2.23637, 48.784747]   

      nbre_pdc               last_modified  
2388         1  2023-03-03T17:39:50.555000  
2407         1  2023-03-03T17:39:50.555000  
2408         1  2023-03-03T17:39:50.555000  
2409         1  2023-03-03T17:39:50.555000  
2410         1  2023-03-03T1

- le principal écart est lié à la présence de l'historique des modifications.
- l'écart résiduel (1068 lignes) est à traiter de façon spécifique -> cf exemples d'écarts ci-dessous

In [28]:
id_pdc = Iindex.merging([champ('id_pdc_itinerance'), champ('last_modified')], 'id_pdc')
ecart = id_pdc.coupling(root)
print('lignes en écart : ',len(ecart))

c = Counter(id_pdc.codec).most_common(5)
id_pdc.reindex()
print('les 5 pdc avec le plus de lignes : \n', c, '\n')

lignes en écart :  1068
les 5 pdc avec le plus de lignes : 
 [(('FR55CP84000', '2023-03-03T17:39:50.555000'), 50), (('Non concerné', '2023-03-03T17:37:07.612000'), 34), (('FR55CP92140', '2023-03-03T17:39:50.555000'), 29), (('FR55CP78140', '2023-03-03T17:39:50.555000'), 28), (('FR55CP83310', '2023-03-03T17:39:50.555000'), 23)] 



In [29]:
data.loc[ecart,['id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY', 'nbre_pdc', 'last_modified']]

id_pdc_itinerance id_station_itinerance  \
3773   FRP07E6748200312        FRP07P67482003   
3772   FRP07E6748200312        FRP07P67482003   
2437        FR55CP84310           FR55CP84310   
2279        FR55CP84310           FR55CP84310   
2540        FR55CP84310           FR55CP84310   
...                 ...                   ...   
8945   FRP07E6748200334        FRP07P67482003   
7890   FRP07E6748200335        FRP07P67482003   
7882   FRP07E6748200335        FRP07P67482003   
21803    FRS87EMB872412        FRS87PMB872412   
21791    FRS87EMB872412        FRS87PMB872412   

                                     nom_station          coordonneesXY  \
3773                          STRASBOURG - Wodli   [7.73762, 48.587002]   
3772                          STRASBOURG - Wodli   [7.73762, 48.587002]   
2437                AVENUE JEAN MONNET - MORIERE  [4.899412, 43.940189]   
2279             PLACE EMMANUEL VITRIA - MORIERE   [4.90471, 43.942662]   
2540             PLACE EMMANUEL VITRIA - MORIERE   [4.90471, 43.942662]   
...                                          ...                    ...   
8945                          STRASBOURG - Wodli   [7.73762, 48.587002]   
7890                          STRASBOURG - Wodli   [7.73762, 48.587002]   
7882                          STRASBOURG - Wodli   [7.73762, 48.587002]   
21803  SAINT-PRIEST-TAURION - ESPLANADE DU STADE  [1.330306, 45.875523]   
21791  SAINT-PRIEST-TAURION - ESPLANADE DU STADE  [1.330306, 45.875523]   

       nbre_pdc               last_modified  
3773          1  2023-03-03T17:39:32.426000  
3772         41  2023-03-03T17:39:32.426000  
2437          1  2023-03-03T17:39:50.555000  
2279          1  2023-03-03T17:39:50.555000  
2540          1  2023-03-03T17:39:50.555000  
...         ...                         ...  
8945          1  2023-03-03T17:39:32.426000  
7890          1  2023-03-03T17:39:32.426000  
7882         41  2023-03-03T17:39:32.426000  
21803         2  2023-03-03T17:38:51.190000  
21791         1  2023-03-03T17:38:51.190000  

[1068 rows x 6 columns]

- l'élimination des doublons (en conservant la date de modification la plus résente) permet d'avoir une ligne par point de charge.

In [30]:
data2 = data.sort_values(by='last_modified').drop_duplicates('id_pdc_itinerance', keep='last').reset_index(drop=True)
print('data2 : ', len(data2))

data2 :  42457


In [31]:
idxs2 = Ilist.obj(data2)
print('idxs (len, lenlidx, sumcodec) : ', len(idxs2), len(idxs2.idxlen), sum(idxs2.idxlen))

idxs (len, lenlidx, sumcodec) :  42457 50 193355


In [32]:
champ = idxs2.nindex
root = Iindex.ext(list(range(len(idxs))))
print("Couplage entre le champ 'id_pdc_itinerance' et la liste des lignes : ", champ('id_pdc_itinerance').iscoupled(root))

Couplage entre le champ 'id_pdc_itinerance' et la liste des lignes :  False


## Représentation des stations

- deux champs sont candidats pour identifier les stations : 'id_station_itinerance', 'nom_station'
- Le champ le plus approprié est 'id_station_itinerance' (plus grand nombre de valeurs différentes)
- une autre solution pourrait être de considérer qu'une station est identifiée par le couple 'id_station_itinerance', 'nom_station' mais ceci conduit à doublonner une station pour laquelle une erreur de nom a été commise (cf exemple ci-dessous).

In [ ]:
print('le nom dérive de la station : ', champ('id_station_itinerance').isderived(champ('nom_station')), '\n')

ecart_nom = champ('id_station_itinerance').coupling(champ('nom_station'))
c_nom = Counter(champ('id_station_itinerance').codec).most_common(5)
champ('id_station_itinerance').reindex()

print('nombre de lignes en écart : ',len(ecart_nom))
print('les 5 stations avec le plus de noms : \n', c_nom, '\n')

In [ ]:
print('stations avec plusieurs noms : \n')
data2.loc[ecart_nom,['id_pdc_itinerance', 'id_station_itinerance', 'nom_station', 'coordonneesXY']]

## localisation des stations
Les stations sont localisées au travers d'une coordonnée géographique et une adresse complète cette localisation. De même que pour la dénomination des stations, deux questions peuvent être posées :
- peut-on avoir plusieurs stations avec la même coordonnée ?
- peut-on avoir plusieurs coordonnées pour une même station ?

L'analyse ci-dessous montre que des erreurs d'identification de stations conduisent à répondre oui à ces deux questions.

In [ ]:
ecart_XY = champ('id_station_itinerance').coupling(champ('coordonneesXY'))
c_XY = Counter(champ('id_station_itinerance').codec).most_common(5)
champ('id_station_itinerance').reindex()

print('nombre de lignes en écart : ',len(ecart_XY))

print('les 5 stations avec le plus de positions différentes : \n', c_XY, '\n')

In [ ]:
print('stations avec plusieurs coordonnées : \n')
data2.loc[ecart_XY,['id_station_itinerance', 'nom_station', 'coordonneesXY']]

In [ ]:
ecart_stat = champ('coordonneesXY').coupling(champ('id_station_itinerance'))
c_stat = Counter(champ('coordonneesXY').codec).most_common(5)
champ('coordonneesXY').reindex()

print('nombre de lignes en écart : ',len(ecart_stat))

print('les 5 positions avec le plus grand nombre de stations associées : \n', c_stat, '\n')

In [ ]:
print('positions avec plusieurs stations associées : \n')
data2.loc[ecart_stat,['coordonneesXY', 'id_station_itinerance', 'nom_station']]

-----------------------
# Annexe : Initialisation du modèle de données 'mermaid'

In [1]:
from base64 import b64encode
from IPython.display import Image, display

def mermaid(graph):
  display(Image(url="https://mermaid.ink/img/" + b64encode(graph.encode("ascii")).decode("ascii")))

In [2]:
# modèle de données suivant l'application 'mermaid' 
irve = """ 
erDiagram
    AMENAGEUR ||..|{ STATION : amenage
    AMENAGEUR {
        string nom_amenageur
        string siren_amenageur
        string contact_amenageur 
    }
    OPERATEUR ||..|{ STATION : "exploite pour le compte de l enseigne"
    OPERATEUR {
        string contact_operateur PK "M"
        string nom_operateur 
        string telephone_operateur 
    }
    ENSEIGNE ||..|{ STATION : "heberge"
    ENSEIGNE {
        string nom_enseigne PK "M" 
    }
    STATION {
        string  id_station_itinerance PK "M"
        string  nom_station "M"
        enum    implantation_station "M"
        integer nbre_pdc "M"
        string  condition_acces "M"
        string  horaires "M"
        boolean station_deux_roues "M"
        string  id_station_local
        enum    raccordement
        string  num_pdl
        date    date_mise_en_service 
    }
    LOCALISATION ||--|{ STATION : "localise"
    LOCALISATION {
       array   coordonneesXY PK "M"
       string  adresse_station "M"
       string  code_insee_commune 
    }
    STATION ||--|{ POINT_DE_CHARGE : regroupe
    POINT_DE_CHARGE {
        string id_pdc_itinerance PK "M Root"
        number puissance_nominale "M"
        boolean prise_type_ef "M"
        boolean prise_type_2 "M"
        boolean prise_type_2combo_ccs "M"
        boolean prise_type_chademo "M"
        boolean prise_type_autre "M"
        boolean paiement_acte "M"
        boolean paiement_autre "M"
        boolean reservation "M"
        enum    accessibilite_pmr "M"
        string  restriction_gabarit "M"
        date    date_maj "M"
        string  id_pdc_local
        boolean gratuit
        boolean paiement_cb
        string  tarification
        string  observations
        boolean cable_t2_attache 
    }
"""